sharding-and-replication.md

```diff
@@ -0,0 +1,156 @@
+        layout:
+          shardsCount: 2
+```
+
+Use the following command to run `clickhouse-client` on Clickhouse nodes and create the distributed tables:
+```bash
+kubectl exec -it service/chi-simple-sharded-1-0 -- clickhouse-client
+```
+
+### Plugin config
+
+The plugin has to be configured to write and read that from the global tables:
+
+```yaml
+address: tcp://clickhouse-simple:9000
+spans_table: jaeger_spans_global
+spans_index_table: jaeger_index_global
+operations_table: jaeger_operations
+```
+
+## Replication
+
+Replication as the name suggest automatically replicates the data across multiple Clickhouse nodes.
+It is used to accomplish high availability, load scaling and migration/updates.
+
+The replication uses Zookeeper. Refer to the Clickhouse operator on how to deploy Zookeeper.
+
+Zookeeper allows us to use `ON CLUSTER` to automatically replicate table creation on all nodes.
+So the following command can be run only on a single Clickhouse node:
+
+```sql
+CREATE TABLE IF NOT EXISTS jaeger_spans ON CLUSTER sharded  (
+                                                                timestamp DateTime CODEC(Delta, ZSTD(1)),
+    traceID String CODEC(ZSTD(1)),
+    model String CODEC(ZSTD(3))
+    ) ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_spans', '{replica}')
+    PARTITION BY toDate(timestamp)
+    ORDER BY traceID
+    SETTINGS index_granularity=1024;
+
+CREATE TABLE IF NOT EXISTS jaeger_index ON CLUSTER sharded (
+                                                               timestamp DateTime CODEC(Delta, ZSTD(1)),
+    traceID String CODEC(ZSTD(1)),
+    service LowCardinality(String) CODEC(ZSTD(1)),
+    operation LowCardinality(String) CODEC(ZSTD(1)),
+    durationUs UInt64 CODEC(ZSTD(1)),
+    tags Array(String) CODEC(ZSTD(1)),
+    INDEX idx_tags tags TYPE bloom_filter(0.01) GRANULARITY 64,
+    INDEX idx_duration durationUs TYPE minmax GRANULARITY 1
+    ) ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_index', '{replica}')
+    PARTITION BY toDate(timestamp)
+    ORDER BY (service, -toUnixTimestamp(timestamp))
+    SETTINGS index_granularity=1024;
+
+CREATE MATERIALIZED VIEW IF NOT EXISTS jaeger_operations ON CLUSTER sharded
+ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_operations', '{replica}')
+PARTITION BY toYYYYMM(date) ORDER BY (date, service, operation)
+SETTINGS index_granularity=32
+POPULATE
+AS SELECT
+              toDate(timestamp) AS date,
+  service,
+  operation,
+  count() as count
+   FROM jaeger_index
+   GROUP BY date, service, operation
+
+
+CREATE TABLE jaeger_spans_global ON CLUSTER sharded AS jaeger_spans ENGINE = Distributed(sharded, default, jaeger_spans, rand());
+CREATE TABLE jaeger_index_global ON CLUSTER sharded AS jaeger_index ENGINE = Distributed(sharded, default, jaeger_index, rand());
+CREATE TABLE jaeger_operations_global on CLUSTER sharded AS jaeger_operations ENGINE = Distributed(sharded, default, jaeger_operations, rand());
+```
```

Comment:
1. It's more universal to use '{cluster}' instead of particular name;
2. Add "IF NOT EXISTS" to creation of global tables;
3. Add ';' after creation of jaeger_operations.

```sql
CREATE TABLE IF NOT EXISTS jaeger_spans ON CLUSTER '{cluster}'  (
                                                                timestamp DateTime CODEC(Delta, ZSTD(1)),
    traceID String CODEC(ZSTD(1)),
    model String CODEC(ZSTD(3))
    ) ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_spans', '{replica}')
    PARTITION BY toDate(timestamp)
    ORDER BY traceID
    SETTINGS index_granularity=1024;

CREATE TABLE IF NOT EXISTS jaeger_index ON CLUSTER '{cluster}' (
                                                               timestamp DateTime CODEC(Delta, ZSTD(1)),
    traceID String CODEC(ZSTD(1)),
    service LowCardinality(String) CODEC(ZSTD(1)),
    operation LowCardinality(String) CODEC(ZSTD(1)),
    durationUs UInt64 CODEC(ZSTD(1)),
    tags Array(String) CODEC(ZSTD(1)),
    INDEX idx_tags tags TYPE bloom_filter(0.01) GRANULARITY 64,
    INDEX idx_duration durationUs TYPE minmax GRANULARITY 1
    ) ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_index', '{replica}')
    PARTITION BY toDate(timestamp)
    ORDER BY (service, -toUnixTimestamp(timestamp))
    SETTINGS index_granularity=1024;

CREATE MATERIALIZED VIEW IF NOT EXISTS jaeger_operations ON CLUSTER '{cluster}'
ENGINE ReplicatedMergeTree('/clickhouse/tables/{shard}/jaeger_operations', '{replica}')
PARTITION BY toYYYYMM(date) ORDER BY (date, service, operation)
SETTINGS index_granularity=32
POPULATE
AS SELECT
              toDate(timestamp) AS date,
  service,
  operation,
  count() as count
   FROM jaeger_index
   GROUP BY date, service, operation;


CREATE TABLE IF NOT EXISTS jaeger_spans_global ON CLUSTER '{cluster}' AS jaeger_spans
    ENGINE = Distributed('{cluster}', default, jaeger_spans, rand());
CREATE TABLE IF NOT EXISTS jaeger_index_global ON CLUSTER '{cluster}' AS jaeger_index 
    ENGINE = Distributed('{cluster}', default, jaeger_index, rand());
CREATE TABLE IF NOT EXISTS jaeger_operations_global on CLUSTER '{cluster}' AS jaeger_operations
    ENGINE = Distributed('{cluster}', default, jaeger_operations, rand());
```

Reply:
> It's more universal to use '{cluster}' instead of particular name;

I like it I will test if the substitution works.